<h1>Analyzing Top Cities</h1>

<hr></hr>
<p>Importing Libraries</p>
<ul>
    <li>Numpy and Pandas for Data wrangling and analysis</li>
    <li> Plotly for Data Visulalization</li>
</ul>

In [91]:
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

`Importing Data`

In [92]:
df = pd.read_csv("data.csv")

<div class="alert alert-block alert-warning">
<b>Empty Values:</b> Finding all the empty or dash(-) placed values
</div>

In [93]:
empty_values = df[df.eq("-").any(1)]
empty_values

City  Rank Sunshine hours(City) Cost of a bottle of water(City)  \
6        Fukuoka     7                 2769                           £0.78   
11       Beijing    12                 2671                           £0.26   
12       Bangkok    13                 2624                           £0.22   
13  Buenos Aires    14                 2525                           £0.57   
16       Jakarta    17                 2983                           £0.21   
19        Geneva    20                    -                           £2.62   
22         Cairo    23                 3542                           £0.16   
23        Taipei    24                 1405                           £0.57   
25        Mumbai    26                 2584                           £0.15   
30     Hong Kong    31                 1836                           £0.75   
31      Shanghai    32                 1776                           £0.29   
35     Sao Paulo    36                 2003                           £0.44   
38  Johannesburg    39                 3124                           £0.59   

   Obesity levels(Country)  Life expectancy(years) (Country)  \
6                    4.30%                              83.2   
11                   6.20%                              75.4   
12                  10.00%                              74.1   
13                  28.30%                              75.9   
16                   6.90%                              68.5   
19                  19.50%                              82.6   
22                  32.00%                              70.7   
23                   6.20%                              75.4   
25                   3.90%                              67.3   
30                   6.20%                              75.4   
31                   6.20%                              75.4   
35                  22.10%                              73.9   
38                  28.30%                              56.3   

   Pollution(Index score) (City) Annual avg. hours worked  \
6                              -                     1644   
11                         85.43                        -   
12                         76.64                        -   
13                         52.64                        -   
16                         84.39                        -   
19                         27.25                     1557   
22                         91.74                        -   
23                         49.32                        -   
25                         82.84                        -   
30                         67.46                        -   
31                         77.40                        -   
35                         79.78                        -   
38                         61.83                        -   

    Happiness levels(Country)  Outdoor activities(City)  \
6                        5.87                        35   
11                       5.12                       223   
12                       5.99                       377   
13                       5.97                       246   
16                       5.28                       114   
19                       7.56                        44   
22                       4.15                       323   
23                       5.12                       134   
25                       3.57                       187   
30                       5.51                       277   
31                       5.12                       108   
35                       6.37                       158   
38                       4.81                       194   

    Number of take out places(City) Cost of a monthly gym membership(City)  
6                               539                                 £55.87  
11                              261                                 £38.62  
12                             1796                                 £50.03  
13                

`Replacing dash with NaN values`

In [94]:
df.replace("-", np.nan, inplace=True)
df.head(10)

City  Rank Sunshine hours(City) Cost of a bottle of water(City)  \
0   Amsterdam     1                 1858                           £1.92   
1      Sydney     2                 2636                           £1.48   
2      Vienna     3                 1884                           £1.94   
3   Stockholm     4                 1821                           £1.72   
4  Copenhagen     5                 1630                           £2.19   
5    Helsinki     6                 1662                           £1.60   
6     Fukuoka     7                 2769                           £0.78   
7      Berlin     8                 1626                           £1.55   
8   Barcelona     9                 2591                           £1.19   
9   Vancouver    10                 1938                           £1.08   

  Obesity levels(Country)  Life expectancy(years) (Country)  \
0                  20.40%                              81.2   
1                  29.00%                              82.1   
2                  20.10%                              81.0   
3                  20.60%                              81.8   
4                  19.70%                              79.8   
5                  22.20%                              80.4   
6                   4.30%                              83.2   
7                  22.30%                              80.6   
8                  23.80%                              82.2   
9                  29.40%                              81.7   

  Pollution(Index score) (City) Annual avg. hours worked  \
0                         30.93                     1434   
1                         26.86                     1712   
2                         17.33                     1501   
3                         19.63                     1452   
4                         21.24                     1380   
5                         13.08                     1540   
6                           NaN                     1644   
7                         39.41                     1386   
8                         65.19                     1686   
9                         24.26                     1670   

   Happiness levels(Country)  Outdoor activities(City)  \
0                       7.44                       422   
1                       7.22                       406   
2                       7.29                       132   
3                       7.35                       129   
4                       7.64                       154   
5                       7.80                       113   
6                       5.87                        35   
7                       7.07                       254   
8                       6.40                       585   
9                       7.23                       218   

   Number of take out places(City) Cost of a monthly gym membership(City)  
0                             1048                                 £34.90  
1                             1103                                 £41.66  
2                             1008                                 £25.74  
3                              598                                 £37.31  
4                              523                                 £32.53  
5                              309                                 £35.23  
6                              539                                 £55.87  
7                             1729                                 £26.11  
8                             2344                                 £37.80  
9                              788                                 £31.04

### Data Wrangling

Dealing with NaN values :-
- There are only three columns with NaN values i.e.
    - Annual avg. hours worked
    - Sunshine hours(City)
    - Pollution(Index score) (City)

- In this approach, I will replace the NaN values with column mean values.
    

*Calculating mean of the columns*

In [95]:
working_hours_mean = pd.to_numeric(df["Annual avg. hours worked"]).mean()
sunshine_hours_mean = pd.to_numeric(df["Sunshine hours(City)"]).mean()
pollution_index_mean = pd.to_numeric(df["Pollution(Index score) (City)"]).mean()
print(working_hours_mean, sunshine_hours_mean, pollution_index_mean)


1672.909090909091 2245.860465116279 51.12232558139535


*Replacing NaN values with Column mean values (rounding off the mean)*

In [96]:
df["Annual avg. hours worked"].replace(np.nan, round(working_hours_mean), inplace=True)
df["Sunshine hours(City)"].replace(np.nan, round(sunshine_hours_mean), inplace=True)
df["Pollution(Index score) (City)"].replace(np.nan, round(pollution_index_mean), inplace=True)

In [18]:
df.isnull().values.any()

False

<div class="alert alert-block alert-success">
No NaN values left in the Dataset!
</div>

Working hours is usually considered Annually, therefore `I will convert Annual avg. hours worked to Weekly avg. hours worked`

In [97]:
df["Annual avg. hours worked"] = round(pd.to_numeric(df["Annual avg. hours worked"]) / 52)
df["Annual avg. hours worked"] = df["Annual avg. hours worked"].astype("int64")
df.rename(columns={"Annual avg. hours worked": "Weekly avg. hours worked"}, inplace=True)

In [43]:
df.head(10)

City  Rank Sunshine hours(City) Cost of a bottle of water(City)  \
0   Amsterdam     1                 1858                           £1.92   
1      Sydney     2                 2636                           £1.48   
2      Vienna     3                 1884                           £1.94   
3   Stockholm     4                 1821                           £1.72   
4  Copenhagen     5                 1630                           £2.19   
5    Helsinki     6                 1662                           £1.60   
6     Fukuoka     7                 2769                           £0.78   
7      Berlin     8                 1626                           £1.55   
8   Barcelona     9                 2591                           £1.19   
9   Vancouver    10                 1938                           £1.08   

  Obesity levels(Country)  Life expectancy(years) (Country)  \
0                  20.40%                              81.2   
1                  29.00%                              82.1   
2                  20.10%                              81.0   
3                  20.60%                              81.8   
4                  19.70%                              79.8   
5                  22.20%                              80.4   
6                   4.30%                              83.2   
7                  22.30%                              80.6   
8                  23.80%                              82.2   
9                  29.40%                              81.7   

  Pollution(Index score) (City)  Weekly avg. hours worked  \
0                         30.93                        28   
1                         26.86                        33   
2                         17.33                        29   
3                         19.63                        28   
4                         21.24                        27   
5                         13.08                        30   
6                            51                        32   
7                         39.41                        27   
8                         65.19                        32   
9                         24.26                        32   

   Happiness levels(Country)  Outdoor activities(City)  \
0                       7.44                       422   
1                       7.22                       406   
2                       7.29                       132   
3                       7.35                       129   
4                       7.64                       154   
5                       7.80                       113   
6                       5.87                        35   
7                       7.07                       254   
8                       6.40                       585   
9                       7.23                       218   

   Number of take out places(City) Cost of a monthly gym membership(City)  
0                             1048                                 £34.90  
1                             1103                                 £41.66  
2                             1008                                 £25.74  
3                              598                                 £37.31  
4                              523                                 £32.53  
5                              309                                 £35.23  
6                              539                                 £55.87  
7                             1729                                 £26.11  
8                             2344                                 £37.80  
9                              788                                 £31.04

### Data Visualization

*Let's check if there's some proper relationship between city rankings and Happiness Level using Line Graph*

In [99]:
fig1 = px.line(df, x='Rank', y='Happiness levels(Country)', hover_data=['City'])
fig1.update_layout(title_text = "LineGraph for Happiness Levels based on Rankings", yaxis_title = "Happiness Levels", xaxis_title="City Rank")
fig1.show()

<div class="alert alert-block alert-info">
<b>Observation:</b> Happiness Level goes up and down with Ranking, therefore, there isn't any proper relationship.
</div>

*Let's make a Bar Chart for Top 10 cleanest cities*

In [88]:
df['Pollution(Index score) (City)'] = df["Pollution(Index score) (City)"].astype(float)
top_clean_cities = df[["Pollution(Index score) (City)", "City"]].sort_values(
                                                        by='Pollution(Index score) (City)', axis=0).head(10)
fig2 = px.bar(top_clean_cities, x="City", y="Pollution(Index score) (City)")
fig2.update_layout(title_text = "BarChart For Least Polluted Cities", yaxis_title = "Pollution Level", xaxis_title="City")
fig2.show()

<div class="alert alert-block alert-info">
<b>Observation:</b> Helsinki is th cleanest city with pollution index of 13.08 followed by Zurich and Vienna.   
</div>

*Let's observe the relationship between pollution levels and life Expectancy*

In [85]:
fig3 = px.scatter(df, x='Pollution(Index score) (City)', y='Life expectancy(years) (Country)', trendline='ols', hover_data=['City'])
fig3.update_layout(title_text = "ScatterPlot for Life Expectancy w.r.t Pollution", yaxis_title = "Life Epectancy (years)", xaxis_title="Pollution Levels")
fig3.show()

<div class="alert alert-block alert-info">
<b>Observation:</b> Life Expectancy tends to decrease with increase in pollution levels. 
</div>

<div class="alert alert-block alert-success">
Johanessburg is an Outlier with very low Life expectancy as compared other cities with similar Pollution Index.
</div>

#### Analyzing Data by Working Hours

`Histogram Based on Weekly working hours`

In [83]:
working_hours_df = df[["Weekly avg. hours worked", 'City']]
fig4 = px.histogram(working_hours_df,x='Weekly avg. hours worked', hover_data=['City'])
fig4.update_layout(title_text = "Histogram For Weekly Working Hours", yaxis_title = "Number of Cities", xaxis_title="Weekly working hours")
fig4.show()

<div class="alert alert-block alert-info">
<b>Observation:</b> 20 cities are working for 32-33 hours weekly while 6 are working for 26-27 and 38-41 working hours . 
</div>

*Let's Separate Cities with Highest and Lowest working hours*

In [40]:
hard_working_cities_df = df[df["Weekly avg. hours worked"] >= 38]
easy_working_cities_df = df[df["Weekly avg. hours worked"] <= 27]

`A Bar Graph showing Happiness levels for both extremes of working hours`

In [77]:
from plotly.subplots import make_subplots

fig = make_subplots(rows=1, cols=2)

fig.add_trace(
    go.Bar(x=hard_working_cities_df['City'], y=hard_working_cities_df['Happiness levels(Country)'], name = "Highest working hours"),
    row=1, col=1
)

fig.add_trace(
    go.Bar(x=easy_working_cities_df['City'], y=easy_working_cities_df['Happiness levels(Country)'],name = "Lowest working hours"),
    row=1, col=2
)

fig.update_layout(height=600, width=800, title_text="Happines levels for Cities with highest and lowest working hours",yaxis=dict(
            range=[0, 8]), yaxis_title = "Happiness Levels", xaxis_title = "Cities")
fig.show()

<div class="alert alert-block alert-info">
<b>Observation:</b>
    <ul>
        <li> Seoul, Moscow and Mexico City have Highest working hours. </li>
        <li> Copenhagen, Berlin and Frankfurt have Lowest working hours. </li>
        <li> Cities with less working hours are Happier than those with more working hours. </li>
    </ul>
</div>